#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [71]:
df.head(20)

Invoice ID Branch       City Customer type  Gender  \
0   750-67-8428      A     Yangon        Member  Female   
1   226-31-3081      C  Naypyitaw        Normal  Female   
2   631-41-3108      A     Yangon        Normal    Male   
3   123-19-1176      A     Yangon        Member    Male   
4   373-73-7910      A     Yangon        Normal    Male   
5   699-14-3026      C  Naypyitaw        Normal    Male   
6   355-53-5943      A     Yangon        Member  Female   
7   315-22-5665      C  Naypyitaw        Normal  Female   
8   665-32-9167      A     Yangon        Member  Female   
9   692-92-5582      B   Mandalay        Member  Female   
10  351-62-0822      B   Mandalay        Member  Female   
11  529-56-3974      B   Mandalay        Member    Male   
12  365-64-0515      A     Yangon        Normal  Female   
13  252-56-2699      A     Yangon        Normal    Male   
14  829-34-3910      A     Yangon        Normal  Female   
15  299-46-1805      B   Mandalay        Member  Female   
16  656-95-9349      A     Yangon        Member  Female   
17  765-26-6951      A     Yangon        Normal    Male   
18  329-62-1586      A     Yangon        Normal    Male   
19  319-50-3348      B   Mandalay        Normal  Female   

              Product line  Unit price  Quantity   Tax 5%     Total  \
0        Health and beauty       74.69         7  26.1415  548.9715   
1   Electronic accessories       15.28         5   3.8200   80.2200   
2       Home and lifestyle       46.33         7  16.2155  340.5255   
3        Health and beauty       58.22         8  23.2880  489.0480   
4        Sports and travel       86.31         7  30.2085  634.3785   
5   Electronic accessories       85.39         7  29.8865  627.6165   
6   Electronic accessories       68.84         6  20.6520  433.6920   
7       Home and lifestyle       73.56        10  36.7800  772.3800   
8        Health and beauty       36.26         2   3.6260   76.1460   
9       Food and beverages       54.84         3   8.2260  172.7460   
10     Fashion accessories       14.48         4   2.8960   60.8160   
11  Electronic accessories       25.51         4   5.1020  107.1420   
12  Electronic accessories       46.95         5  11.7375  246.4875   
13      Food and beverages       43.19        10  21.5950  453.4950   
14       Health and beauty       71.38        10  35.6900  749.4900   
15       Sports and travel       93.72         6  28.1160  590.4360   
16       Health and beauty       68.93         7  24.1255  506.6355   
17       Sports and travel       72.61         6  21.7830  457.4430   
18      Food and beverages       54.67         3   8.2005  172.2105   
19      Home and lifestyle       40.30         2   4.0300   84.6300   

         Date   Time      Payment    cogs  gross margin percentage  \
0    1/5/2019  13:08      Ewallet  522.83                 4.761905   
1    3/8/2019  10:29         Cash   76.40                 4.761905   
2    3/3/2019  13:23  Credit card  324.31                 4.761905   
3   1/27/2019  20:33      Ewallet  465.76                 4.761905   
4    2/8/2019  10:37      Ewallet  604.17                 4.761905   
5   3/25/2019  18:30      Ewallet  597.73                 4.761905   
6   2/25/2019  14:36      Ewallet  413.04                 4.761905   
7   2/24/2019  11:38      Ewallet  735.60                 4.761905   
8   1/10/2019  17:15  Credit card   72.52                 4.761905   
9   2/20/2019  13:27  Credit card  164.52                 4.761905   
10   2/6/2019  18:07      Ewallet   57.92                 4.761905   
11   3/9/2019  17:03         Cash  102.04                 4.761905   
12  2/12/2019  10:25      Ewallet  234.75                 4.761905   
13   2/7/2019  16:48      Ewallet  431.90                 4.761905   
14  3/29/2019  19:21         Cash  713.80                 4.761905   
15  1/15/2019  16:19         Cash  562.32                 4.761905   
16  3/11/2019  11:03  Credit card  482.51                 4.761905   
17   1/1/2019

In [2]:
df = pd.read_csv('/Users/s1ndly/Data Analytics/data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [17]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

Branch C (Naypyitaw)has the highest gross income at $5265.1765

However, Branch A (Yangon) shows the highest gross margin %, which is a major indicator of profitability.
As data indicates, Branch C has higher cogs, which, in  turn, impacts the gross margin %. Therefore, my suggestion would be to consider cities Yangon and Mandalay. as there have highers returns. 

In [19]:
df.groupby(['Branch', 'City'])['gross income'].sum()

Branch  City     
A       Yangon       5057.1605
B       Mandalay     5057.0320
C       Naypyitaw    5265.1765
Name: gross income, dtype: float64

In [20]:
df.groupby(['Branch', 'City'])['cogs'].sum()

Branch  City     
A       Yangon       101143.21
B       Mandalay     101140.64
C       Naypyitaw    105303.53
Name: cogs, dtype: float64

In [21]:
df.groupby(['Branch', 'City'])['gross margin percentage'].sum()

Branch  City     
A       Yangon       1619.047619
B       Mandalay     1580.952381
C       Naypyitaw    1561.904762
Name: gross margin percentage, dtype: float64

In [29]:
df.pivot_table(index='City', columns='Branch', values='gross income', aggfunc='sum', fill_value='None')

Branch             A         B          C
City                                     
Mandalay        None  5057.032       None
Naypyitaw       None      None  5265.1765
Yangon     5057.1605      None       None

In [30]:
df.pivot_table(index='City', columns='Branch', values='cogs', aggfunc='sum', fill_value='None')

Branch             A          B          C
City                                      
Mandalay        None  101140.64       None
Naypyitaw       None       None  105303.53
Yangon     101143.21       None       None

In [31]:
df.pivot_table(index='City', columns='Branch', values='gross margin percentage', aggfunc='sum', fill_value='None')

Branch               A            B            C
City                                            
Mandalay          None  1580.952381         None
Naypyitaw         None         None  1561.904762
Yangon     1619.047619         None         None

In [26]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [45]:
invoices_by_category = df.pivot_table(values='Invoice ID', index='Branch', columns='Product line', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [70]:
invoices_by_category.groupby('Branch')['Electronic accessories'].max()[:1]

Branch
A    60
Name: Electronic accessories, dtype: int64

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


In [133]:
df_woman = df[df.Gender == 'Female']
df_woman.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
6  355-53-5943      A     Yangon        Member  Female   
7  315-22-5665      C  Naypyitaw        Normal  Female   
8  665-32-9167      A     Yangon        Member  Female   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
6  Electronic accessories       68.84         6  20.6520  433.6920  2/25/2019   
7      Home and lifestyle       73.56        10  36.7800  772.3800  2/24/2019   
8       Health and beauty       36.26         2   3.6260   76.1460  1/10/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
6  14:36      Ewallet  413.04                 4.761905       20.6520     5.8  
7  11:38      Ewallet  735.60                 4.761905       36.7800     8.0  
8  17:15  Credit card   72.52                 4.761905        3.6260     7.2

In [152]:
df_man = df[df.Gender == 'Male']
df_man.head()

Invoice ID Branch       City Customer type Gender  \
2   631-41-3108      A     Yangon        Normal   Male   
3   123-19-1176      A     Yangon        Member   Male   
4   373-73-7910      A     Yangon        Normal   Male   
5   699-14-3026      C  Naypyitaw        Normal   Male   
11  529-56-3974      B   Mandalay        Member   Male   

              Product line  Unit price  Quantity   Tax 5%     Total  \
2       Home and lifestyle       46.33         7  16.2155  340.5255   
3        Health and beauty       58.22         8  23.2880  489.0480   
4        Sports and travel       86.31         7  30.2085  634.3785   
5   Electronic accessories       85.39         7  29.8865  627.6165   
11  Electronic accessories       25.51         4   5.1020  107.1420   

         Date   Time      Payment    cogs  gross margin percentage  \
2    3/3/2019  13:23  Credit card  324.31                 4.761905   
3   1/27/2019  20:33      Ewallet  465.76                 4.761905   
4    2/8/2019  10:37      Ewallet  604.17                 4.761905   
5   3/25/2019  18:30      Ewallet  597.73                 4.761905   
11   3/9/2019  17:03         Cash  102.04                 4.761905   

    gross income  Rating  
2        16.2155     7.4  
3        23.2880     8.4  
4        30.2085     5.3  
5        29.8865     4.1  
11        5.1020     6.8

In [157]:
df_man_with_hour.pivot_table(values='Time', index='Product line', columns='City', aggfunc=['max', 'min'])

max                       min                 
City                   Mandalay Naypyitaw Yangon Mandalay Naypyitaw Yangon
Product line                                                              
Electronic accessories    20:57     20:37  20:54    10:38     10:02  10:46
Fashion accessories       20:52     20:54  20:06    11:45     10:00  10:00
Food and beverages        20:03     20:37  20:35    10:37     10:02  10:16
Health and beauty         20:03     20:44  20:35    10:11     10:30  10:12
Home and lifestyle        20:38     20:59  20:07    10:26     10:11  10:16
Sports and travel         20:08     19:49  20:19    10:11     10:17  10:03

In [145]:
df_woman_with_hour.pivot_table(values='Time', index='Product line', columns='City', aggfunc=['max', 'min'])

max                       min                 
City                   Mandalay Naypyitaw Yangon Mandalay Naypyitaw Yangon
Product line                                                              
Electronic accessories    20:51     20:51  20:59    10:17     10:23  10:17
Fashion accessories       20:42     20:30  18:20    10:33     10:06  10:11
Food and beverages        20:55     20:39  20:48    10:38     10:17  10:00
Health and beauty         20:43     19:57  20:36    10:38     10:18  10:05
Home and lifestyle        20:24     20:50  19:48    10:39     10:13  10:09
Sports and travel         19:35     19:54  20:34    10:01     10:01  10:18

In [148]:
df_woman_with_hour.pivot_table(values='Quantity', index='Product line', columns='City', aggfunc=['count'])

count                 
City                   Mandalay Naypyitaw Yangon
Product line                                    
Electronic accessories       28        28     28
Fashion accessories          33        35     28
Food and beverages           29        38     23
Health and beauty            20        23     21
Home and lifestyle           22        25     32
Sports and travel            30        29     29

In [158]:
df_man_with_hour.pivot_table(values='Quantity', index='Product line', columns='City', aggfunc=['count'])

count                 
City                   Mandalay Naypyitaw Yangon
Product line                                    
Electronic accessories       27        27     32
Fashion accessories          29        30     23
Food and beverages           21        28     35
Health and beauty            33        29     26
Home and lifestyle           28        20     33
Sports and travel            32        16     30

In [177]:
time = '11:28'



get_hour(time)
get_hour = lambda t: int(t[:2])

df_woman['Hour'] = df_woman.Time.apply(lambda t: int(t[:2]))
df_woman_with_hour = df_woman.copy()

df_man['Hour'] = df_man.Time.apply(lambda t: int(t[:2]))
df_man_with_hour = df_man.copy()


df['Hour'] = df.Time.apply(lambda t: int(t[:2]))
df = df.copy()

/var/folders/18/v6tsqf5s6tn983tw5hm40jb00000gn/T/ipykernel_8728/2526048007.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_woman['Hour'] = df_woman.Time.apply(lambda t: int(t[:2]))
/var/folders/18/v6tsqf5s6tn983tw5hm40jb00000gn/T/ipykernel_8728/2526048007.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_man['Hour'] = df_man.Time.apply(lambda t: int(t[:2]))


In [225]:
date = '3/21/2019'

get_date = lambda d: int(d[:1])


df['Month'] = df.Date.apply(lambda d: int(d[:1]))
df = df.copy()

In [231]:
df.tail(10)

Invoice ID Branch       City Customer type  Gender            Product line  Unit price  Quantity   Tax 5%      Total       Date   Time      Payment    cogs  gross margin percentage  gross income  Rating  Hour  Month
990  886-18-2897      A     Yangon        Normal  Female      Food and beverages       56.56         5  14.1400   296.9400  3/22/2019  19:06  Credit card  282.80                 4.761905       14.1400     4.5    19      3
991  602-16-6955      B   Mandalay        Normal  Female       Sports and travel       76.60        10  38.3000   804.3000  1/24/2019  18:10      Ewallet  766.00                 4.761905       38.3000     6.0    18      1
992  745-74-0715      A     Yangon        Normal    Male  Electronic accessories       58.03         2   5.8030   121.8630  3/10/2019  20:46      Ewallet  116.06                 4.761905        5.8030     8.8    20      3
993  690-01-6631      B   Mandalay        Normal    Male     Fashion accessories       17.49        10   8.7450   183.6450  2/22/2019  18:35      Ewallet  174.90                 4.761905        8.7450     6.6    18      2
994  652-49-6720      C  Naypyitaw        Member  Female  Electronic accessories       60.95         1   3.0475    63.9975  2/18/2019  11:40      Ewallet   60.95                 4.761905        3.0475     5.9    11      2
995  233-67-5758      C  Naypyitaw        Normal    Male       Health and beauty       40.35         1   2.0175    42.3675  1/29/2019  13:46      Ewallet   40.35                 4.761905        2.0175     6.2    13      1
996  303-96-2227      B   Mandalay        Normal  Female      Home and lifestyle       97.38        10  48.6900  1022.4900   3/2/2019  17:16      Ewallet  973.80                 4.761905       48.6900     4.4    17      3
997  727-02-1313      A     Yangon        Member    Male      Food and beverages       31.84         1   1.5920    33.4320   2/9/2019  13:22         Cash   31.84                 4.761905        1.5920     7.7    13      2
998  347-56-2442      A     Yangon        Normal    Male      Home and lifestyle       65.82         1   3.2910    69.1110  2/22/2019  15:33         Cash   65.82                 4.761905        3.2910     4.1    15      2
999  849-09-3807      A     Yangon        Member  Female     Fashion accessories       88.34         7  30.9190   649.2990  2/18/2019  13:28         Cash  618.38                 4.761905       30.9190     6.6    13      2

In [227]:
df_man_with_hour.head()

Invoice ID Branch       City Customer type Gender            Product line  Unit price  Quantity   Tax 5%     Total       Date   Time      Payment    cogs  gross margin percentage  gross income  Rating  Hour
2   631-41-3108      A     Yangon        Normal   Male      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019  13:23  Credit card  324.31                 4.761905       16.2155     7.4    13
3   123-19-1176      A     Yangon        Member   Male       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019  20:33      Ewallet  465.76                 4.761905       23.2880     8.4    20
4   373-73-7910      A     Yangon        Normal   Male       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019  10:37      Ewallet  604.17                 4.761905       30.2085     5.3    10
5   699-14-3026      C  Naypyitaw        Normal   Male  Electronic accessories       85.39         7  29.8865  627.6165  3/25/2019  18:30      Ewallet  597.73                 4.761905       29.8865     4.1    18
11  529-56-3974      B   Mandalay        Member   Male  Electronic accessories       25.51         4   5.1020  107.1420   3/9/2019  17:03         Cash  102.04                 4.761905        5.1020     6.8    17

In [109]:
df_woman2.pivot_table(values='Quantity', index='Product line', columns='City', aggfunc=['count', 'max', 'min'])

count                       max                   \
City                   Mandalay Naypyitaw Yangon Mandalay Naypyitaw Yangon   
Product line                                                                 
Electronic accessories       28        28     28       10        10     10   
Fashion accessories          33        35     28       10        10     10   
Food and beverages           29        38     23       10        10     10   
Health and beauty            20        23     21       10        10     10   
Home and lifestyle           22        25     32       10        10     10   
Sports and travel            30        29     29       10        10     10   

                            min                   
City                   Mandalay Naypyitaw Yangon  
Product line                                      
Electronic accessories        1         1      1  
Fashion accessories           1         1      1  
Food and beverages            1         1      1  
Health and beauty             1         1      1  
Home and lifestyle            1         1      1  
Sports and travel             1         1      1

In [100]:
df_pivoted = df.pivot_table(values=['Quantity'], index=['Product line', 'Gender'], columns=['City'], aggfunc=['count', 'max', 'min', 'mean'])
df_pivoted

count                       max            \
                              Quantity                  Quantity             
City                          Mandalay Naypyitaw Yangon Mandalay Naypyitaw   
Product line           Gender                                                
Electronic accessories Female       28        28     28       10        10   
                       Male         27        27     32       10        10   
Fashion accessories    Female       33        35     28       10        10   
                       Male         29        30     23       10        10   
Food and beverages     Female       29        38     23       10        10   
                       Male         21        28     35       10        10   
Health and beauty      Female       20        23     21       10        10   
                       Male         33        29     26       10        10   
Home and lifestyle     Female       22        25     32       10        10   
                       Male         28        20     33       10        10   
Sports and travel      Female       30        29     29       10        10   
                       Male         32        16     30       10         9   

                                          min                       mean  \
                                     Quantity                   Quantity   
City                          Yangon Mandalay Naypyitaw Yangon  Mandalay   
Product line           Gender                                              
Electronic accessories Female     10        1         1      1  5.571429   
                       Male       10        1         1      1  5.925926   
Fashion accessories    Female     10        1         1      1  5.363636   
                       Male       10        1         1      1  4.137931   
Food and beverages     Female     10        1         1      1  5.827586   
                       Male       10        1         1      1  4.809524   
Health and beauty      Female     10        1         1      1  5.600000   
                       Male       10        1         1      1  6.303030   
Home and lifestyle     Female     10        1         1      1  6.681818   
                       Male       10        2         1      1  5.285714   
Sports and travel      Female     10        1         1      1  5.000000   
                       Male       10        1         1      1  5.375000   

                                                   
                                                   
City                          Naypyitaw    Yangon  
Product line           Gender                      
Electronic accessories Female  5.857143  6.000000  
                       Male    6.259259  4.812500  
Fashion accessories    Female  5.485714  5.750000  
                       Male    5.000000  4.434783  
Food and beverages     Female  6.078947  4.956522  
                       Male    4.928571  5.685714  
Health and beauty      Female  5.217391  5.285714  
                       Male    5.413793  5.615385  
Home and lifestyle     Female  6.080000  6.218750  
                       Male    4.650000  5.212121  
Sports and travel      Female  6.551724  5.379310  
                       Male    4.687500  5.900000

In [102]:
df_pivoted2 = df.pivot_table(values=['Unit price'], index=['Product line', 'Gender'], columns=['City'], aggfunc=['max', 'min', 'mean'])
df_pivoted2

max                         min  \
                              Unit price                  Unit price   
City                            Mandalay Naypyitaw Yangon   Mandalay   
Product line           Gender                                          
Electronic accessories Female      99.73     99.69  99.56      10.59   
                       Male        91.56     96.82  99.55      18.93   
Fashion accessories    Female      97.61     99.30  99.10      14.48   
                       Male        99.89     99.82  98.98      17.49   
Food and beverages     Female      99.69     99.42  99.60      17.63   
                       Male        93.40     99.79  99.78      18.08   
Health and beauty      Female      99.71     99.19  95.95      13.50   
                       Male        99.16     99.96  99.83      10.75   
Home and lifestyle     Female      97.38     89.25  96.52      11.28   
                       Male        99.92     97.50  97.94      13.59   
Sports and travel      Female      98.13     98.80  98.40      11.85   
                       Male        99.96     99.24  98.09      13.69   

                                                     mean             \
                                               Unit price              
City                          Naypyitaw Yangon   Mandalay  Naypyitaw   
Product line           Gender                                          
Electronic accessories Female     12.45  15.69  47.677857  54.031429   
                       Male       11.81  10.56  52.111111  57.653333   
Fashion accessories    Female     10.18  12.09  50.135758  57.203429   
                       Male       12.78  17.94  60.201379  62.690667   
Food and beverages     Female     14.87  10.96  60.691379  59.173947   
                       Male       17.04  10.13  48.426190  54.693214   
Health and beauty      Female     10.16  10.69  50.206000  51.824783   
                       Male       15.37  10.08  63.021818  59.260000   
Home and lifestyle     Female     10.53  11.43  60.632727  53.424000   
                       Male       13.98  18.28  51.492143  55.472000   
Sports and travel      Female     14.36  12.34  56.256333  54.860345   
                       Male       10.17  12.76  62.885937  55.555000   

                                          
                                          
City                              Yangon  
Product line           Gender             
Electronic accessories Female  54.445357  
                       Male    55.243750  
Fashion accessories    Female  59.785357  
                       Male    52.878261  
Food and beverages     Female  60.846522  
                       Male    51.115714  
Health and beauty      Female  51.124762  
                       Male    48.842692  
Home and lifestyle     Female  57.895000  
                       Male    53.858485  
Sports and travel      Female  51.920000  
                       Male    59.177667

In [178]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

   Hour  
0    13  
1    10  
2    13  
3    20  
4    10

In [196]:
df_products = df.pivot_table(index=['Product line', 'Customer type', 'Gender'], columns=['City'], values=['Quantity'], aggfunc=['count', 'max', 'mean']).round(2)
df_products

count                   \
                                            Quantity                    
City                                        Mandalay Naypyitaw Yangon   
Product line           Customer type Gender                             
Electronic accessories Member        Female       14         9     12   
                                     Male         13        11     19   
                       Normal        Female       14        19     16   
                                     Male         14        16     13   
Fashion accessories    Member        Female       16        19     12   
                                     Male         16        13     10   
                       Normal        Female       17        16     16   
                                     Male         13        17     13   
Food and beverages     Member        Female       18        23     10   
                                     Male         11        13     19   
                       Normal        Female       11        15     13   
                                     Male         10        15     16   
Health and beauty      Member        Female        7        11     12   
                                     Male         19        14     10   
                       Normal        Female       13        12      9   
                                     Male         14        15     16   
Home and lifestyle     Member        Female       12        19     16   
                                     Male          9         9     18   
                       Normal        Female       10         6     16   
                                     Male         19        11     15   
Sports and travel      Member        Female       18        15     18   
                                     Male         12        13     11   
                       Normal        Female       12        14     11   
                                     Male         20         3     19   

                                                 max                   \
                                            Quantity                    
City                                        Mandalay Naypyitaw Yangon   
Product line           Customer type Gender                             
Electronic accessories Member        Female       10        10     10   
                                     Male         10        10     10   
                       Normal        Female       10        10     10   
                                     Male         10        10      9   
Fashion accessories    Member        Female       10        10     10   
                                     Male         10        10     10   
                       Normal        Female       10        10     10   
                                     Male         10        10     10   
Food and beverages     Member        Female       10        10      9   
                                     Male          9        10     10   
                       Normal        Female        8         9     10   
                                     Male         10        10     10   
Health and beauty      Member        Female        9        10      9   
                                     Male         10        10     10   
                       Normal        Female       10         9     10   
                                     Male         10        10      9   
Home and lifestyle     Member        Female        9        10     10   
                                     Male         10         8     10   
                       Normal        Female       10        10     10   
                                     Male          9        10      9   
Sports and travel      Member        Female        9        10     10   
                                     Male         10         7     10   
                       Normal        Female       10        10      7   
                               

In [217]:
df_naypyitaw = df[(df['City'] == 'Naypyitaw') & (df['Product line'] == 'Electronic accessories') & (df['Gender'] == 'Female')]
unit_price_comparison = df_naypyitaw.groupby('Customer type')['Unit price'].mean()
print(unit_price_comparison)


Customer type
Member    62.961111
Normal    49.801579
Name: Unit price, dtype: float64


In [220]:
product_analysis = df.groupby(['Customer type', 'Product line']).agg({'Quantity': 'mean', 'Total': 'mean'})
print(product_analysis)


                                      Quantity       Total
Customer type Product line                                
Member        Electronic accessories  5.500000  314.083269
              Fashion accessories     5.104651  306.092581
              Food and beverages      5.382979  333.591702
              Health and beauty       5.863014  353.849856
              Home and lifestyle      5.903614  337.084663
              Sports and travel       5.666667  324.532190
Normal        Electronic accessories  5.891304  324.337353
              Fashion accessories     5.032609  304.151446
              Food and beverages      5.575000  309.840300
              Health and beauty       5.392405  295.730373
              Home and lifestyle      5.467532  336.154364
              Sports and travel       5.405063  340.361089


In [211]:
df[df['Product line'] == 'Electronic accessories'].groupby(['City', 'Customer type', 'Gender'])[['Unit price']].agg(['mean', 'max', 'min', 'std']).round(2)

Unit price                     
                                     mean    max    min    std
City      Customer type Gender                                
Mandalay  Member        Female      44.19  93.96  10.59  29.94
                        Male        53.07  91.56  18.93  25.31
          Normal        Female      51.16  99.73  13.78  28.80
                        Male        51.22  87.08  21.58  23.99
Naypyitaw Member        Female      62.96  98.84  30.20  22.31
                        Male        66.18  96.82  12.05  29.57
          Normal        Female      49.80  99.69  12.45  25.31
                        Male        51.79  96.37  11.81  27.79
Yangon    Member        Female      55.32  94.64  17.42  25.37
                        Male        54.98  93.78  11.94  28.95
          Normal        Female      53.79  99.56  15.69  25.54
                        Male        55.63  99.55  10.56  30.47

In [213]:
df.groupby(['City', 'Product line', 'Customer type','Gender'])['Unit price'].agg(['mean']).round(2)

mean
City      Product line           Customer type Gender       
Mandalay  Electronic accessories Member        Female  44.19
                                               Male    53.07
                                 Normal        Female  51.16
                                               Male    51.22
          Fashion accessories    Member        Female  46.98
                                               Male    57.38
                                 Normal        Female  53.11
                                               Male    63.68
          Food and beverages     Member        Female  57.66
                                               Male    53.13
                                 Normal        Female  65.65
                                               Male    43.25
          Health and beauty      Member        Female  45.37
                                               Male    63.88
                                 Normal        Female  52.81
                                               Male    61.86
          Home and lifestyle     Member        Female  61.29
                                               Male    43.91
                                 Normal        Female  59.84
                                               Male    55.08
          Sports and travel      Member        Female  56.56
                                               Male    71.00
                                 Normal        Female  55.80
                                               Male    58.02
Naypyitaw Electronic accessories Member        Female  62.96
                                               Male    66.18
                                 Normal        Female  49.80
                                               Male    51.79
          Fashion accessories    Member        Female  66.24
                                               Male    68.39
                                 Normal        Female  46.48
                                               Male    58.33
          Food and beverages     Member        Female  61.30
                                               Male    54.59
                                 Normal        Female  55.91
                                               Male    54.78
          Health and beauty      Member        Female  51.94
                                               Male    62.84
                                 Normal        Female  51.72
                                               Male    55.92
          Home and lifestyle     Member        Female  53.52
                                               Male    56.89
                                 Normal        Female  53.11
                                               Male    54.31
          Sports and travel      Member        Female  52.39
                                               Male    47.52
                                 Normal        Female  57.50
                                               Male    90.38
Yangon    Electronic accessories Member        Female  55.32
                                               Male    54.98
                                 Normal        Female  53.79
                                               Male    55.63
          Fashion accessories    Member        Female  52.82
                                               Male    43.16
                                 Normal        Female  65.01
                                               Male    60.36
          Food and beverages     Member        Female  66.90
                                               Male    55.18
                                 Normal        Female  56.19
                                               Male    46.28
          Health and beauty      Member        Female  50.43
                                               Male    50.55
                                 Normal        Female  52.05
                                               Male    47.78
          Home and lifestyle  

In [203]:

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)


In [246]:
df_products_time = df.pivot_table(index=['Product line', 'Customer type'], columns=['City'], values=['Time'], aggfunc=['max', 'min'])
df_products_time

max                       min                 
                                         Time                      Time                 
City                                 Mandalay Naypyitaw Yangon Mandalay Naypyitaw Yangon
Product line           Customer type                                                    
Electronic accessories Member           20:32     20:51  20:54    10:17     11:21  10:17
                       Normal           20:57     20:36  20:59    10:34     10:02  10:18
Fashion accessories    Member           20:13     20:08  19:30    11:16     10:36  10:00
                       Normal           20:52     20:54  20:06    10:33     10:00  10:12
Food and beverages     Member           20:55     20:39  20:35    10:37     10:17  10:00
                       Normal           20:43     20:37  20:48    10:48     10:02  10:16
Health and beauty      Member           20:24     20:44  20:33    10:11     10:30  11:03
                       Normal           20:43     20:05  20:36    10:25     10:18  10:05
Home and lifestyle     Member           19:50     20:59  20:07    10:53     10:11  10:10
                       Normal           20:38     20:50  19:48    10:26     10:55  10:09
Sports and travel      Member           19:51     19:49  20:34    11:02     10:01  10:37
                       Normal           20:08     19:54  20:19    10:01     10:02  10:03

In [222]:
df[df['Customer type'] == 'Member'].pivot_table(index=['Product line', 'Customer type'], columns=['City'], values=['Quantity'], aggfunc=['sum'], margins=True)

sum                       
                                     Quantity                       
City                                 Mandalay Naypyitaw Yangon   All
Product line           Customer type                                
Electronic accessories Member             147       120    162   429
Fashion accessories    Member             143       163    133   439
Food and beverages     Member             160       193    153   506
Health and beauty      Member             172       127    129   428
Home and lifestyle     Member             137       143    210   490
Sports and travel      Member             165       151    177   493
All                                       924       897    964  2785

In [249]:
df_time_analysis = df.groupby('Customer type')['Hour'].agg(['mean', 'median', 'max', 'min']).round(2)
df_time_analysis

mean  median  max  min
Customer type                         
Member         14.97    15.0   20   10
Normal         14.85    15.0   20   10

In [267]:
gender_analysis = df[df['Product line'] == 'Fashion accessories'].groupby(['Gender', 'City'])[['Quantity', 'Unit price', 'Total']].agg(['mean']).round(2)
pd.DataFrame(gender_analysis)


Quantity Unit price   Total
                     mean       mean    mean
Gender City                                 
Female Mandalay      5.36      50.14  275.20
       Naypyitaw     5.49      57.20  329.13
       Yangon        5.75      59.79  351.29
Male   Mandalay      4.14      60.20  252.82
       Naypyitaw     5.00      62.69  334.68
       Yangon        4.43      52.88  282.45

1. Normal-type customers are buying higher quantity of Electronic Accessories than Members in Naypyitaw city (and even aacross all the cities). The mean price for Members is 62.96$, while for Normal is $49.80, meaning that Normal customers are bulk-buying products in this category, whereas Members are buying lower quantities, but paying a premium price.

Naypyitaw -> Electronic Accessories: Female - 9 (Member), 19 (Normal)
Min price - 30.20 (member), 12.45 (normal)

If Members have higher purchasing power, then the business could utilise it in a way of expanding its product range to offer more luxury items and adjusting marketing campaign to target wealthier individuals.

2. Do Normal customers tend to shop earlier in the day compared to Members?
Member	14.97
Normal	14.85

This confirms the hypothesis. Therefore, the business may offer promotions targeted at Normal cusotmers in order to increase sales volume.

3. Do female customers prefer buying more items in the 'Fashion accessories' category compared to male customers?

Female

Mandalay	5.36	
Naypyitaw	5.49
Yangon	5.75

Male	
Mandalay	4.14
Naypyitaw	5.00
Yangon	4.43

Action: Sign up for more collaborationd with celebrities or bloggers to increase attention